Improved dnode allocation

ofaaland opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

<!---
Documentation on ZFS Buildbot options can be found at
https://github.com/zfsonlinux/zfs/wiki/Buildbot-Options
-->

### Description
<!--- Describe your changes in detail -->
Refactor dmu_hold_impl() and () to fix incorrectly handled races, improve readability, and possibly improve performance.

### Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
There are at least two incorrectly-handled cases in the existing code:
dmu_hold_impl() called by two threads concurrently, with the same object id but different slot sizes, can return a successful hold to both threads.  One of the threads will then win the race to take dn_struct_lock() and allocate the dnode, and it may allocate it with the wrong slot count, overwriting another dnode.

The chunk-handling code at the top of dmu_object_alloc_dnsize() can fail to detect the end of the chunk with dnodesize > 1, resulting in all threads operating in the same chunk without correct coordination.

### How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
<!--- If your change is a performance enhancement, please provide benchmarks here. -->
Minor local testing.  Additional testing yet to be done.

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Performance enhancement (non-breaking change which improves efficiency)
- [x] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the ZFS on Linux code style requirements.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [x] All commit messages are properly formatted and contain `Signed-off-by`.
- [ ] Change has been approved by a ZFS on Linux member.

behlendorf commented on include/sys/dnode.h:
```diff
@@ -463,6 +463,11 @@ typedef struct dnode_stats {
+	/*
+	 * Number of times dmu_object_alloc*() reached the end of the existing
+	 * object ID chunk and advanced to a new one.
+	 */
```
The comment should be above the `kstat_named_t`.

ofaaland replied:
fixed

behlendorf commented on module/zfs/dmu_object.c:
```diff
@@ -158,6 +160,13 @@ dmu_object_alloc_dnsize(objset_t *os, dmu_object_type_t ot, int blocksize,
+		 * ID assigned to us.  The value afterwards is the object ID
+		 * assigned to whoever wants to do an allocation next.
+		 */
+		object =  atomic_add_64_nv(cpuobj, dn_slots) - dn_slots;
```
nit: extra space after `=`.

ofaaland replied:
fixed

behlendorf commented on cmd/zdb/zdb.c:
```diff
@@ -1989,7 +1989,7 @@ dump_object(objset_t *os, uint64_t object, int verbosity, int *print_header)
 	zdb_nicenum(doi.doi_physical_blocks_512 << 9, asize);
 	zdb_nicenum(doi.doi_bonus_size, bonus_size);
-	zdb_nicenum(doi.doi_dnodesize, dnsize);
+	zdb_nicenum(doi.doi_dnodesize/512, dnsize);
```
You could use DNODE_MIN_SIZE here.

ofaaland replied:
I was actually thinking I should just alter zdb to compute the unused slot count itself, rather than doing so in the test.  Then this patch wouldn't be needed.

I personally would rather see dnode slots than dnode size, but figured the output here should match the dnodesize property.  What do you think?

behlendorf replied:
That would probably be better.  Upon further reflection I think we should leave this reporting the size in bytes which is more consistent with the property.

ofaaland replied:
fixed

behlendorf commented on tests/zfs-tests/tests/functional/features/large_dnode/large_dnode_009_pos.ksh:
```diff
@@ -0,0 +1,70 @@
+
+log_must wait
+
+log_must zpool export $TESTPOOL
```
Running `zdb -d` to verify the object count would also be a good check.

ofaaland replied:
fixed

sempervictus commented:
The race described in the commit log sounds terrifying. Must be a rare case - we run a "fair amount" of ops and data over our cinder pools, haven't hit this yet AFAIK. Would a scrub catch a race-win-written dnode error?

ofaaland commented:
> The race described in the commit log sounds terrifying. Must be a rare case - we run a "fair amount" of ops and data over our cinder pools, haven't hit this yet AFAIK.
It requires dnodesize != legacy (the default value), and lots of concurrent object creates.  We saw it ourselves when using Lustre which often creates many objects at the same time.

> Would a scrub catch a race-win-written dnode error?
Yes.

behlendorf commented:
@ofaaland can you fold the fix for #6559 in to this PR.  The DNODE_MUST_BE_ALLOCATED case fell out for free in the refactoring, but the DNODE_MUST_BE_FREE case was converted to an ASSERT.  Let's instead handle that as an error as was previously the case.  Here's (most of) a proposed patch.

```diff
diff --git a/module/zfs/dnode.c b/module/zfs/dnode.c
index 2c0bd95..60ed1e2 100644
--- a/module/zfs/dnode.c
+++ b/module/zfs/dnode.c
@@ -1374,9 +1374,16 @@ dnode_hold_impl(objset_t *os, uint64_t object, int flag, 
                }
 
                mutex_enter(&dn->dn_mtx);
+               if (!refcount_is_zero(&dn->dn_holds)) {
+                       DNODE_STAT_BUMP(dnode_hold_free_refcount);
+                       mutex_exit(&dn->dn_mtx);
+                       dnode_slots_rele(dnc, idx, slots);
+                       dbuf_rele(db, FTAG);
+                       return (SET_ERROR(EEXIST));
+               }
+
                dnode_set_slots(dnc, idx + 1, slots - 1, DN_SLOT_INTERIOR);
                DNODE_STAT_BUMP(dnode_hold_free_hits);
-               ASSERT(refcount_is_zero(&dn->dn_holds));
        } else {
                dbuf_rele(db, FTAG);
                return (SET_ERROR(EINVAL));
```

ofaaland commented:
> @ofaaland can you fold the fix for #6559 in to this PR

done

behlendorf commented on module/zfs/dnode.c:
```diff
@@ -1326,7 +1326,15 @@ dnode_hold_impl(objset_t *os, uint64_t object, int flag, int slots,
+			 * we acquired the lock.
+			 */
+			if (DN_SLOT_IS_PTR(dnh->dnh_dnode))
+				dn = dnh->dnh_dnode;
```
What do you think about adding a `dnode_hold_alloc_lock_misses` kstat here as a counterpart to `dnode_hold_free_lock_misses`.  It could be insightful to know how often this actually happens.

ofaaland replied:
I'll do that.

behlendorf commented on tests/zfs-tests/tests/functional/features/large_dnode/large_dnode_009_pos.ksh:
```diff
@@ -0,0 +1,70 @@
+	else
+		do_unlink="-f $((RANDOM % 1024))"
+	fi
+	log_must eval "xattrtest -R -r -y -x 1 $do_unlink -p $dir &"
```
This end up being pretty chatty, you might want to redirect everything to /dev/null.

behlendorf commented:
@ofaaland the other additional fixes look correct to me.  The next time you update this I'd suggest rebasing on master and squashing the PR for a final round of review and testing.

ofaaland commented:
> @ofaaland the other additional fixes look correct to me. The next time you update this I'd suggest rebasing on master and squashing the PR for a final round of review and testing.

OK, will do.  I'm working on testing with random dnode sizes now.

ofaaland commented:
@behlendorf , I've made the changes you suggested, squashed, and rebased.  The only other change is to fix the percent empty verification in the test shell script.  Created appx 2M files in the course of running tests 008 and 009 with zfs_mknode() altered to choose a random dnode size and did not encounter any issue there.

behlendorf commented on module/zfs/dnode.c:
```diff
@@ -988,14 +1003,88 @@ dnode_move(void *buf, void *newbuf, size_t size, void *arg)
+static int
+dnode_slots_tryenter(dnode_children_t *children, int idx, int slots)
+{
+	dnode_handle_t *dnh;
```
nit: we could pull this declaration down in to the `for` loop like in `dnode_slots_rele`

ofaaland replied:
fixed

behlendorf commented:
Performance looks good.  There's no impact to the per-thread create rate and an approximate 5% improvement when using many threads/  That's on top of the previous multi-threaded allocation improvements.  The following graph shows the sustained rate of creation of 4M million files are a variety of thread counts.

![pr6564](https://user-images.githubusercontent.com/148917/29991079-6791a986-8f36-11e7-9899-6202d903bbaf.png)

MT - The [multi-threaded allocator](dbeb8796996b4a342f7de2b3eb4ea5b86ac260f9) is applied.
PR6564 - The patch from this PR is applied.

sempervictus commented:
The MT allocator link is dead... Pretty chart though :)

kernelOfTruth commented:
MT allocator (for reference):
https://github.com/zfsonlinux/zfs/pull/6117
https://github.com/zfsonlinux/zfs/commit/dbeb8796996b4a342f7de2b3eb4ea5b86ac260f9

ofaaland commented:
Refreshed the patch.  Fixed the nit re: dnode_slots_tryenter() and some mistakes in the commit messsage, and rebased on master.  No other changes.

ofaaland commented:
@behlendorf Testing on our Lustre cluster was successful.
